src/cart/crt_hg.c

```diff
@@ -858,6 +1255,115 @@ crt_rpc_handler_common(hg_handle_t hg_hdl)
 	D_ASSERT(proc != NULL);
 	opc = rpc_tmp.crp_req_hdr.cch_opc;
 
+	if (crt_gdata.cg_na_plugin == CRT_NA_OFI_PSM2) {
+
+		/* Determine whether this came from cleint or server in order to
+		 * determine which hash table to use
+		 */
+		rpc_tmp_pub = &rpc_tmp.crp_pub;
+
+		rc = crt_req_src_rank_get(rpc_tmp_pub, &hdr_src_rank);
+		if (rc != 0) {
+			D_ERROR("failed to retrieve rpc src rank, rc: %d.\n",
+						rc);
+			D_GOTO(out, hg_ret = -DER_NONEXIST);
+		}
+
+		/** Search for the recieved hg_addr in the ha hash table. If
+		 *  not found, then insert.
+		 */
+
+		if (hdr_src_rank == CRT_NO_RANK) {
+			ha_hash_table_rwlock_ptr =
+					&crt_ctx->cc_ha_hash_table_rwlock;
+			ha_hash_table_ptr = &crt_ctx->cc_ha_hash_table;
+			client_id = rpc_tmp.crp_req_hdr.cch_clid;
+		} else {
+			ha_hash_table_rwlock_ptr =
+				&crt_ctx->cc_ha_server_hash_table_rwlock;
+			ha_hash_table_ptr = &crt_ctx->cc_ha_server_hash_table;
+			client_id = (uint64_t)hdr_src_rank;
+		}
+		D_RWLOCK_WRLOCK(ha_hash_table_rwlock_ptr);
+		halink = d_hash_rec_find(ha_hash_table_ptr, (void *)&client_id,
+				sizeof(client_id));
+		if (halink != NULL) {
+			/*Traverse the list and add new entry to tail
+  			 *if not found. 
```

Comment:
(style)  trailing whitespace
(style)  code indent should use tabs where possible
(style)  please, no space before tabs